File: app_mvp/src/main/java/com/example/jqorz/mvptest/BaseMvpFragment.java
package com.example.jqorz.mvptest;


import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.fragment.app.Fragment;

import com.example.jqorz.mvptest.lib.MvpView;
import com.example.jqorz.mvptest.lib.delegate.fragment.FragmentDelegate;
import com.example.jqorz.mvptest.lib.delegate.fragment.FragmentDelegateImp;
import com.example.jqorz.mvptest.lib.delegate.fragment.FragmentMvpDelegateCallback;
import com.example.jqorz.mvptest.lib.presenters.MvpPresenter;


/**
 * 使用MVP的Fragment
 */
public abstract class BaseMvpFragment<P extends MvpPresenter<V>, V extends MvpView> extends Fragment implements
        FragmentMvpDelegateCallback<P, V>, MvpView {
    private FragmentDelegate mFragmentDelegate;
    private P mPresenter;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        mFragmentDelegate = new FragmentDelegateImp<>(this);
        mFragmentDelegate.onCreateView();

        View view = inflater.inflate(getLayoutResId(), container, false);
        init();

        return view;
    }

    @Override
    public void onPause() {
        super.onPause();
        mFragmentDelegate.onPause();
    }

    @Override
    public void onResume() {
        super.onResume();
        mFragmentDelegate.onResume();
    }

    @Override
    public void onStop() {
        super.onStop();
        mFragmentDelegate.onStop();
    }

    protected abstract void init();

    protected abstract int getLayoutResId();

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mFragmentDelegate.onDestroyView();
    }

    @Override
    public void setPresenter() {

    }

    //暴露一个创建的方法用于创建presenter
    protected abstract P CreatePresenter();

    @Override
    //这个方法由MvpInternalDelegate 调用BaseDelegateCallback 来创建presenter
    public P createPresenter() {
        mPresenter = CreatePresenter();
        return mPresenter;
    }

    @Override
    public P getPresenter() {
        return mPresenter;
    }

    @Override
    public V getMvpView() {
        return (V) this;
    }
}

File: app_mvp/src/main/java/com/example/jqorz/mvptest/lib_kotlin/BaseMvpFragment.kt
package com.jqorz.tallybook.base.mvp

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.jqorz.tallybook.base.mvp.delegate.fragment.FragmentDelegate
import com.jqorz.tallybook.base.mvp.delegate.fragment.FragmentDelegateImp
import com.jqorz.tallybook.base.mvp.delegate.fragment.FragmentMvpDelegateCallback
import com.jqorz.tallybook.base.mvp.presenters.MvpPresenter

/**
 * 使用MVP的Fragment
 */
abstract class BaseMvpFragment<P : MvpPresenter<V>?, V : MvpView> : Fragment(), FragmentMvpDelegateCallback<P, V>, MvpView {
    private lateinit var mFragmentDelegate: FragmentDelegate
    private var mPresenter: P? = null
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        mFragmentDelegate = FragmentDelegateImp(this).apply {
            onCreateView()
        }
        val view = inflater.inflate(layoutResId, container, false)
        init()
        return view
    }

    override fun onPause() {
        super.onPause()
        mFragmentDelegate.onPause()
    }

    override fun onResume() {
        super.onResume()
        mFragmentDelegate.onResume()
    }

    override fun onStop() {
        super.onStop()
        mFragmentDelegate.onStop()
    }

    protected abstract fun init()
    protected abstract val layoutResId: Int
    override fun onDestroyView() {
        super.onDestroyView()
        mFragmentDelegate.onDestroyView()
    }

    override fun setPresenter() {}

    //暴露一个创建的方法用于创建presenter
    protected abstract fun createThePresenter(): P

    //这个方法由MvpInternalDelegate 调用BaseDelegateCallback 来创建presenter
    override fun createPresenter(): P {
        mPresenter = createThePresenter()
        return mPresenter!!
    }

    override fun getPresenter(): P {
        return mPresenter!!
    }

    override fun getMvpView(): V {
        return this as V
    }
}